### PR TITLE
WIP: Better copyto

### PR DESCRIPTION
**Pull request by bkamins**
PR following the discussion in #170.

What does it change:
* `append!` uses `resize!` and `copyto!` which will make it faster, but when it fails it does not clean up
* `copy!` will allow `src` to be non-categorical; `dest` is array or a view; in this implementation levels are retained and also if `dest` is ordered `copy!` might fail
* `copyto!`:
    1. if `dest` is ordered do not allow adding new levels
    2. if `dest` is not ordered only add levels that are present in `src` in copied range (not all levels in `src` - this makes sure that if `src` is categorical and not categorical holding the same values as `src` we get the same result)

**EDIT**
The general idea of the design is that if `x` is categorical and `y` is a non-categorical then:
```
copyto!(x,a,y,b,n)
```
should be exactly the same as
```
copyto!(x,a,categorical(y),b,n)
```

### CODE REVIEW

**Comment by bkamins** [comment]
I am going through the tests (hopefully I will update all today and push relevant changes) and here are two additional comments:
* `copy!` indeed should have a different implementation - indeed it should copy levels and moreover it should have a totally separate implementation (I will add it) as `Future.copy!([1,2,3],[5,6]) == [5,6]` and we do not resize catarrays with `copy!` right now;
* `copyto!` behavior that I propose is inconsistent with `vcat` in terms of ordering of levels (we could change `vcat` implementation to use `copyto!` internally to make it consistent); this should be discussed again I think; also the usage of `copyto!` in DataFrames.jl assumes that level ordering is retained so if we use what `copyto!` now proposes the pattern to populate a bigger vector using `copyto!` should be:
    1. create an empty categorical vector
    2. populate its levels as needed
    3. populate the values after levels are populated

File: src/array.jl [hunk 1]
```diff
@@ -419,43 +420,49 @@ function copyto!(dest::CatArrOrSub{T, N}, dstart::Integer,
...
-    else # More work to do: preserve some values (and therefore index)
-        levels!(dpool, newlevels)
-
-        indexmap = indexin(index(spool), index(dpool))
-
-        @inbounds for i = 0:(n-1)
-            x = srefs[sstart+i]
-            drefs[dstart+i] = x > 0 ? indexmap[x] : 0
+    if isordered(dest)
+        remap = Vector{eltype(drefs)}(undef, length(levels(src)))
```
Comment (nalimilan) [inline src/array.jl]:
```suggestion
        remap = similar(levels(src), eltype(drefs))
```

File: src/array.jl [hunk 2]
```diff
@@ -419,43 +420,49 @@ function copyto!(dest::CatArrOrSub{T, N}, dstart::Integer,
...
+            s = srefs[sstart+i]
+            if eltype(src) >: Missing && s == 0
+                if eltype(dest) >: Missing
+                    drefs[dstart+i] = 0
+                else
+                    throw(MissingException("cannot copy array with missing values to an array with element type $T"))
+                end
+            else
+                if seen[s]
+                    drefs[dstart+i] = remap[s]
```
Comment (nalimilan) [inline src/array.jl]: Couldn't you just use `remap[s] == -1` as a sentinel?

File: src/array.jl [hunk 3]
```diff
@@ -419,43 +420,49 @@ function copyto!(dest::CatArrOrSub{T, N}, dstart::Integer,
...
+                else
+                    seen[s] = true
+                    rm = dpool.invindex[s]
```
Comment (nalimilan) [inline src/array.jl]: AFAICT this doesn't work, and better avoid using internal fields:
```suggestion
                    rm = index(dpool)[s]
```
(The call to `index` would better be hoisted out of the loop.)

File: src/array.jl [hunk 4]
```diff
@@ -419,43 +420,49 @@ function copyto!(dest::CatArrOrSub{T, N}, dstart::Integer,
...
+                    remap[s] = rm
+                    drefs[dstart+i] = rm
+                end
+            end
+        end
+    else
```
Comment (nalimilan) [inline src/array.jl]: Are you sure we need to duplicate all of this code for performance? Branch prediction is usually quite efficient, so even if you have an `if` inside the loop I suspect it won't make a big difference. And often LLVM will do the hoisting automatically.

File: src/array.jl [hunk 5]
```diff
@@ -464,11 +471,10 @@ end
...
     copyto!(dest, 1, src, 1, length(src))
 
-copyto!(dest::CatArrOrSub, dstart::Integer, src::CatArrOrSub) =
-    copyto!(dest, dstart, src, 1, length(src))
-
 @static if VERSION >= v"0.7.0-DEV.3208"
     using Future
+    Future.copy!(dest::CatArrOrSub, src::AbstractArray) =
+        copyto!(dest, 1, src, 1, length(src))
     Future.copy!(dest::CatArrOrSub, src::CatArrOrSub) =
```
Comment (nalimilan) [inline src/array.jl]: This method isn't needed now AFAICT.

File: src/array.jl [hunk 6]
```diff
@@ -694,14 +700,12 @@ function Base.push!(A::CategoricalVector, item)
...
-function Base.append!(A::CategoricalVector, B::CatArrOrSub)
-    levels!(A, union(levels(A), levels(B)))
+function Base.append!(A::CategoricalVector, B::AbstractArray)
     len = length(A)
     len2 = length(B)
     resize!(A.refs, len + len2)
-    for i = 1:len2
-        A[len + i] = B[i]
-    end
+    # As in Base, A will be left modified if it is not possible to copy B to A
```
Comment (nalimilan) [inline src/array.jl]: I wonder why Base doesn't call `resize!` in case of failure. That sounds trivial, undoing the operation is not as hard as in `copyto!`. But better stick to what Base does, and maybe file an issue in Julia.

File: test/13_arraycommon.jl [hunk 7]
```diff
@@ -442,6 +461,32 @@ end
         end
     end
 
+    @testset "copyto! and copy! corner cases" begin
+        x = categorical([1:255;1:255], true)
+        y = categorical([1, 1000], true)
+        @test copyto!(x, 1, y, 1, 1)[1] == y[1]
+        @test_throws ErrorException copyto!(x, 1, y, 1, 1)[2]
```
Comment (nalimilan) [inline test/13_arraycommon.jl]: Better check the length? Actually, better check the whole contents of the result (same for the line above).

File: test/13_arraycommon.jl [hunk 8]
```diff
@@ -442,6 +461,32 @@ end
...
+
+        x = categorical([1,2,3])
+        y = categorical([5,6,7])
```
Comment (nalimilan) [inline test/13_arraycommon.jl]: Better use unsorted values to really test that the order is preserved.

File: test/13_arraycommon.jl [hunk 9]
```diff
@@ -442,6 +461,32 @@ end
...
+        copyto!(x,y)
+        @test levels(x) = [1,2,3,5,6,7]
+
+        @test copy!(x, [1,1,1]) == [1,1,1]
```
Comment (nalimilan) [inline test/13_arraycommon.jl]: Why not also check levels? Also with different unsorted values? Same below.

**Comment by nalimilan** [comment]
Thanks! That inconsistency between `copyto!` and `vcat` is a bit annoying. I'd say the essential property that should hold for `copyto!` is that it's equivalent to calling `setindex!` repeatedly. Currently, `setindex!(::CategoricalArray, ::Any, ::CatValue)` doesn't respect the ordering of levels (i.e. it adds a new level at the end if needed), which means we need `copyto!` to do the same.

But maybe we could change that? It would make sense for `setindex!` to insert the level in the position in which it would have in the result of `mergelevels`, i.e. in the levels of `vcat`. In general, the principle I've put a lot of effort into implementing is that if the two vectors have compatible orderings of levels, we try very hard to merge the levels into a common order (i.e. one in which the relative order of levels is preserved). I think that's right since the order of appearance is really an arbitrary criterion which shouldn't play a role for categorical arrays.

File: src/array.jl (hunk 2 above)
Comment (bkamins) [reply src/array.jl]: This is what I have done first, but allocating `seen` should be cheap, and the problem is that `remap`, in order to be fast, should have the same eltype as `drefs` and they are unsigned.

At least this is what I thought. Now I see that we can use `remap[s] == 0` as sentinel, as we use `0` for missing which is handled in other way anyway. I will change this

File: src/array.jl (hunk 3 above)
Comment (bkamins) [reply src/array.jl]: Thanks - this is the part you know best. I will also move it out from the loop.

File: src/array.jl (hunk 5 above)
Comment (bkamins) [reply src/array.jl]: What do you want to happen if we want to `copy!` a normal array to a categorical array? Throw an error?

File: src/array.jl (hunk 4 above)
Comment (bkamins) [reply src/array.jl]: I would have to do `isordered` check in the innermost part of the loop. I can check if compiler can optimize this out.

File: src/array.jl (hunk 6 above)
Comment (bkamins) [reply src/array.jl]: Maybe the reason is that you would have to do try-catch which will slow down the function? I will open an issue with Base.

**Comment by bkamins** [comment]
>  I've put a lot of effort into implementing is that if the two vectors have compatible orderings of levels, we try very hard to merge the levels into a common order

This is what I see in the code. And that is why I am asking, because the issue is complex (in general - working on those things taught me that things that I initially see as simple are not).

So we have two options:
1. dynamically add levels as needed (what I currently implemented in copyto!)
2. always merge levels

I can implement both of them with equal speed in `copyto!` (of course option 2 with `setindex!` will be slower)

Option 2 is very tempting conceptually. The problem with it is that what we merge might have conflicting orderings (but I guess we should treat it as a corner case and just go with your implementation)?

File: src/array.jl (hunk 4 above)
Comment (nalimilan) [reply src/array.jl]: Since there's already a branch I don't think it will matter, even if the compiler doesn't hoist it.

**Comment by nalimilan** [comment]
If you have the time and motivation, it would be interesting to investigate whether it would be possible to have `setindex!` preserve ordering using `mergelevels`. When orderings conflicts, the fallback is to add levels at the end.

File: src/array.jl (hunk 4 above)
Comment (bkamins) [reply src/array.jl]: but we have to do `isordered` test in every step of the loop. If it is not optimized out then it will be much slower. I will check.

**Comment by bkamins** [comment]
This is what I intended. I really like to have things clean.
However, what I will do, given the raising complexity of this PR is to handle `append!` separately as this is a simple change that will stay anyway.

**Comment by bkamins** [comment]
@nalimilan - sorry to bother you again. I have started working on the order-preserving `setindex!` which led me to the analysis of `mergelevels`, because I have to reimplement it to level updating version (not creating a completely new level).

Can you please explain me the logic behind the following output?

```
julia> CategoricalArrays.mergelevels(false, [1,2,3], [4,2,5,1])
([5, 4, 1, 2, 3], false)
```
I do not understand why you put `4` after `5` in the result.

**Comment by nalimilan** [comment]
No idea. I admit I would have expected `[1, 2, 3, 4, 5]` (due to the order of levels, not to integer sorting). I guess this has to do with the particular solution I found to insert levels while preserving order when possible: AFAIK it works when it's possible to preserve order, but when that's not possible it might not return an optimal choice. Feel free to suggest better solutions if you can find ones, I must say I don't remember all the details now.

**Comment by bkamins** [comment]
I will design such an algorithm, but this is a longer process to make it efficient and consistent for: `setindex!`, `vcat` and `copyto!`. But this will be mildly breaking.

Therefore, because of the complexity we face, I will close this PR and close #170 and open issues/PRs that are split into smaller chunks of changed functionality.

**Comment by bkamins** [comment]
I am thinking about this issue for some time. The first conclusion is that we have to have a list of methods we should hande. This is what I have collected:
* `setindex!` 
* `copyto!`
* `copy!`
* `vcat` (we need to implement a proper method here for CatValue and CatString as they are missing)
* `append!`
* `push!`
* `hcat` (we still have to implement a custom method here as it is missing)
* `hvcat` (we still have to implement a custom method here as it is missing)
* `pushfirst!` (we still have to implement a custom method here as it is missing)
* `insert!` (we still have to implement a custom method here as it is missing)

The key issue is that in general it is impossible to implement `mergelevels` in a consistent way. What I mean by this is that the result must depend on the order of arguments and this is unavoidable.

The general logic for all methods should be:
* check if the operation allows adding new levels
* if it does then add levels
* if it does not then throw an error if the level would have to be added
* perform an operation on a modified target (we know that it allows adding levels)

There are then multiple corner cases but the main is:

> in cases those methods add levels to a vector should they add all levels at once or only levels present in what is added.

Currently with `mergelevels` we add all levels, not only those present. I think it is important to have one guiding princple here that is commonly respected by all the functions, as otherwise users will have a hard time learning the API.

The difference in the approach manifests itself e.g. when doing `setindex!`, `append!`, `copyto!` to an ordered catvector. First of all - they all should be consistent. Now if we have *add all levels* rule then those methods should fail if what we add has some levels that are not supported by the catvector we write to even if the value itself would be accepted. If we have *add needed levels* rule then such operation will work if what we want to add is a level that is accepted.

The thing is that if we do not have common guiding principles then e.g. `vcat(x, y)` and `append!(x, y)` for some data might both work but produce a different result and I think that `vcat(x,y)` and append!(x,y)` should pass equality of results and should pass equality of `levels`.

To understand what I mean have a look at this example (this is one of many possible examples):
```
julia> x = categorical([1,2,3], ordered=true)
3-element CategoricalArray{Int64,1,UInt32}:
 1
 2
 3

julia> y = categorical([4], ordered=true)
1-element CategoricalArray{Int64,1,UInt32}:
 4

julia> levels!(y, [1,4,2])
1-element CategoricalArray{Int64,1,UInt32}:
 4

julia> levels(hcat(x[1], y))
2-element Array{CategoricalValue{Int64,UInt32},1}:
 1
 4

julia> levels(hcat(x[1:1], y))
2-element Array{Int64,1}:
 1
 4

julia> levels(vcat(x[1], y))
2-element Array{CategoricalValue{Int64,UInt32},1}:
 1
 4

julia> levels(vcat(x[1:1], y))
4-element Array{Int64,1}:
 1
 4
 2
 3

julia> levels(append!(x,y))
4-element Array{Int64,1}:
 1
 2
 3
 4
```

So in short - I think we first need a principle we want to follow and then reimplement all the methods from grounds-up to follow it. Unfortunately it seems to be a large task, but - if we want to provide a solid base for the future - I would try to do it.

**Comment by nalimilan** [comment]
Unfortunately I don't think we can fix `cat` at this point, since we would introduce ambiguities with other array types. See https://github.com/JuliaLang/julia/pull/20815. So all the situations where you end up with an `Array{CategoricalValue}` should not exist -- we can't handle this kind of array very well anyway and they don't make a lot of sense.

Regarding whether we should (try to) add all levels or only those actually used, I think we should do the former, because it's not always possible to efficiently check which levels are used (you basically need to make an additional pass over the data). That's one of the reasons why we keep unused levels in the first place (the other being that levels are supposed to have a meaning by themselves, independent from the data).

**Comment by bkamins** [comment]
TODO (from Slack):
also consider this case:
```
using DataFrames
v=DataFrame(:a=>categorical([nothing,"a"]))
append!(v,v)
```
which currently throws an error

**Comment by nalimilan** [comment]
@bkamins Do you know what is still relevant in this PR after we merge #211 and given what we have decided at #99?

**Comment by bkamins** [comment]
I will test it after #211 is merged and tagged.

**Comment by bkamins** [comment]
There is still a problem that:
```
using DataFrames
v=DataFrame(:a=>categorical([nothing,"a"]))
append!(v,v)
```
throws an error.

Also before we close this please confirm that this is intended:
```
julia> x = categorical([1,2,3], ordered=true)
3-element CategoricalArray{Int64,1,UInt32}:
 1
 2
 3

julia> push!(x, 10)
ERROR: cannot add new level 10 since ordered pools cannot be extended implicitly. Use the levels! function to set new levels, or the ordered! function to mark the pool as unordered.

julia> append!(x, [10])
ERROR: MethodError: no method matching append!

julia> append!(x, categorical([10]))
4-element CategoricalArray{Int64,1,UInt32}:
 1
 2
 3
 10
```

**Comment by nalimilan** [comment]
> There is still a problem that:
> 
> ```
> using DataFrames
> v=DataFrame(:a=>categorical([nothing,"a"]))
> append!(v,v)
> ```
> 
> throws an error.

I don't think CategoricalArrays handle `Union` eltypes very well. Probably worth filing a separate issue.

> Also before we close this please confirm that this is intended:
> 
> ```
> julia> x = categorical([1,2,3], ordered=true)
> 3-element CategoricalArray{Int64,1,UInt32}:
>  1
>  2
>  3
> 
> julia> push!(x, 10)
> ERROR: cannot add new level 10 since ordered pools cannot be extended implicitly. Use the levels! function to set new levels, or the ordered! function to mark the pool as unordered.
> 
> julia> append!(x, [10])
> ERROR: MethodError: no method matching append!
> 
> julia> append!(x, categorical([10]))
> 4-element CategoricalArray{Int64,1,UInt32}:
>  1
>  2
>  3
>  10
> ```

The `push!` error is intended (we can discuss that, though). The `append!` error is probably an oversight worth filing another issue (I wish Base would define a fallback based on `resize!` and `setindex!`, but currently it doesn't).

Nothing to change in `copyto!` itself?

**Comment by bkamins** [comment]
If this:
```
julia> x = categorical([1,2,3], ordered=true)
3-element CategoricalArray{Int64,1,UInt32}:
 1
 2
 3

julia> y = categorical([1,2,1000], ordered=true)
3-element CategoricalArray{Int64,1,UInt32}:
 1
 2
 1000

julia> y[3] = 1
1

julia> y
3-element CategoricalArray{Int64,1,UInt32}:
 1
 2
 1

julia> copyto!(x, y)
3-element CategoricalArray{Int64,1,UInt32}:
 1
 2
 1

julia> levels(x)
4-element Array{Int64,1}:
    1
    2
    3
 1000
```
is intended then I think we are OK and this can be closed.

Can I leave opening of the other specific issues you have indicated to you as I feel you understand better what is the core problem with them?

**Comment by bkamins** [comment]
just for a reference continuing the last example, we have:
```
julia> x = categorical([1,2,3], ordered=true)
3-element CategoricalArray{Int64,1,UInt32}:
 1
 2
 3

julia> z = [x; y]
6-element CategoricalArray{Int64,1,UInt32}:
 1
 2
 3
 1
 2
 1

julia> levels(z)
4-element Array{Int64,1}:
    1
    2
    3
 1000
```
so we are consistent.

**Comment by nalimilan** [comment]
> is intended then I think we are OK and this can be closed.
Yes that's intended. At least that's what we agreed upon AFAICT. Good to know we're consistent now!

> Can I leave opening of the other specific issues you have indicated to you as I feel you understand better what is the core problem with them?

Actually I've realized it's covered by https://github.com/JuliaData/CategoricalArrays.jl/issues/170.

**Comment by bkamins** [comment]
OK - so this can be closed then. Right?

**Comment by nalimilan** [comment]
I guess.

I've filed https://github.com/JuliaData/CategoricalArrays.jl/pull/213 to fix the `append!` problem. There are probably many other issues like that, though.
